OpenMRS Atom feed integration

kaapstorm opened the pull request:
This PR adds the ability to poll an OpenMRS Atom feed, and import patient changes.

Tests and testing to follow.

@proteusvacuum and maybe @dannyroberts who might have more context.

A couple of notes:

* There is no obvious (to me) way to skip patient updates that original from CommCare, but it might be that changes made from the API are not shown in the Atom feed. This would make sense considering the Atom feed was originally written specifically for MOTECH (v. 1). A little testing should clear this up.

* There is a limitation that the patient's OpenMRS UUID must be stored in the case's `external_id` property. This is already a requirement for projects that import cases from OpenMRS, so not really an issue. But it would be nice to be able to store it in another case property (like, say, `openmrs_uuid`) and use Case Search to index and find cases. This was considered out of scope, but might be nice to have in the future, if we find we need it.

* The current poll interval is every 10 minutes. For projects where OpenMRS is used in hospitals and clinics and CommCare is used for field visits, it is very unlikely that there would be a race condition where a new patient is added in both systems between polls and before the CommCare case is forwarded to OpenMRS. Currently a race condition will have the following result:
  1. When the Atom feed is polled, and if the case has not yet been forwarded to OpenMRS, then no CommCare case will be found with the patient's UUID as its external ID.
  2. A duplicate case will be created, and its external ID set.
  3. The CommCare case will be forwarded, and if its properties match the same patient, then its external ID will be set to the same UUID.
  4. These cases will co-exist in CommCare with the same external ID until the OpenMRS patient appears in the Atom feed again. At this point a soft_assert will be raised, and I'll have to merge the cases ... unless someone has a better workflow.

* Race conditions can be avoided by using Case Search + something like a [PatientFinder](https://github.com/dimagi/commcare-hq/blob/master/corehq/motech/openmrs/finders.py#L71) if no case's external_id matches an OpenMRS patient's UUID. Again, out of scope at this point, but if we ever encounter a race condition, YAGNI will fall away immediately.

Thoughts and ideas welcome, of course.

kaapstorm committed:
Docs

kaapstorm committed:
Refactor: Uncouple property map from Finder

kaapstorm committed:
Add new OpenmrsRepeater properties

proteusvacuum commented on corehq/motech/utils.py:
```diff
@@ -98,3 +99,9 @@ def pformat_json(data):
         return data
+
+
+def get_commcare_users_by_location(domain_name, location_id):
```
can you use a pattern similar to this so you don't have to pull all web users too?: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/dbaccessors.py#L69-L71 ?

kaapstorm replied:
Thanks. That's much better.

kaapstorm committed:
Tests

kaapstorm committed:
Task to poll Atom feed

kaapstorm committed:
Don't get patients from paused repeaters

kaapstorm committed:
Make timezone-aware

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm commented:
Feature flag: "Enable OpenMRS integration"

kaapstorm commented:
Previously I noted,

> it might be that changes made from the API are not shown in the Atom feed. This would make sense considering the Atom feed was originally written specifically for MOTECH (v. 1). A little testing should clear this up.

"A little testing" showed that patients created by MOTECH do show up in the Atom Feed, and there is, indeed, no way to tell that they came from CommCare, and so their CommCare case is updated with the data from OpenMRS. This is essentially a NOOP.

kaapstorm commented:
@proteusvacuum, finally, this is ready for your eyeballs.

snopoke commented on corehq/motech/openmrs/atom_feed.py:
```diff
@@ -0,0 +1,269 @@
+        repeater.username,
+        repeater.password
+    )
+    last_polled_at = pytz.utc.localize(repeater.atom_feed_last_polled_at)
```
why is this localization necessary?

kaapstorm replied:
The Atom Feed's timestamps are timezone-aware. So when we compare timestamps (in lines 109, 117) this timestamp also needs to be timezone-aware. `atom_feed_last_polled_at` is a UTC timestamp (`datetime.utcnow()`) but the timezone gets dropped by `DateTimeProperty`.

snopoke commented on corehq/motech/openmrs/atom_feed.py:
```diff
@@ -0,0 +1,269 @@
+                ):
+                    if (
+                        not last_polled_at or
+                        get_timestamp(entry, './atom:published') > last_polled_at
```
this condition is the same as above - could make it more readable to extract it to an function?

snopoke commented on corehq/motech/openmrs/atom_feed.py:
```diff
@@ -0,0 +1,269 @@
+            'domain: "{}". case external_id: "{}". repeater: "{}".'.format(
+                repeater.domain, patient_uuid, repeater.get_id
+            )
+        )
```
Is there action required for these alerts? Just wondering what their utility is.

kaapstorm replied:
This particular alert needs action. (I'll check the others.) All those scenarios would be bad: 

* Identifiers are expected to be unique and if they're not then patients are probably duplicated in OpenMRS
* If PatientFinder is configured badly we could end up with duplicates, or worse, false matches where one patient's details get written over by someone else's
* Or race conditions (very unlikely) where cases end up duplicated, including with duplicate `external_id`.

I need to document what to do about this error. It might be a good idea to add a pointer to more documentation in this error message.

snopoke commented on corehq/motech/openmrs/openmrs_config.py:
```diff
@@ -149,3 +155,103 @@ class OpenmrsConfig(DocumentSchema):
+                Cmp(Child(Fields('attributeType'), Fields('uuid')), eq, attr_uuid)
+            ),
+            Fields('value')
+        )
```
what's the benefit of writing it like this vs having jsonpath parse the string version?

kaapstorm replied:
We wrote `Cmp`, but extending the jsonpath parser to support it was going to take more time. I figured it was better just to use it like this for now, and extend the parser as a nice-to-have later.

snopoke commented on corehq/motech/openmrs/tasks.py:
```diff
@@ -252,9 +251,14 @@ def import_patients():
-    pass
+    domains = find_domains_with_toggle_enabled(toggles.OPENMRS_INTEGRATION)
+    for domain in domains:
+        poll_openmrs_atom_feeds(domain)
```
Might be worth making this a task of its own so that a failure in one domain doesn't kill all.

kaapstorm committed:
has_new_entries_since function

kaapstorm committed:
Check if atom_feed_last_polled_at is set

kaapstorm committed:
poll_openmrs_atom_feeds task

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm committed:
Data types and serializers

kaapstorm committed:
ValueSource serializes its own data

kaapstorm committed:
Nit

kaapstorm committed:
Drop redundant serialize function.

kaapstorm committed:
Use ValueSource.deserialize() to get case values

kaapstorm committed:
Set data types for known properties

This preserves current behaviour for serializing OpenMRS datetimes.

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm commented on corehq/motech/openmrs/const.py:
```diff
@@ -51,42 +56,64 @@
+    'endDate': OPENMRS_DATA_TYPE_DATETIME,
 }
+OPENMRS_PROPERTIES = dict(chain(
+    PERSON_PROPERTIES.items(),
```
`chain` is iterating dict.items

kaapstorm committed:
Stickler

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm committed:
Refactor get_property_map

kaapstorm committed:
Finder must deserialize patient value

kaapstorm committed:
Attributes must match simple and Concept values

kaapstorm committed:
Be explicit about taking one value

kaapstorm committed:
Useful serializers

kaapstorm committed:
Update tests

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm committed:
Stickler

kaapstorm committed:
Temporary measure

This is necessary for Py3, but the "jsonpath_rw" library it extends is
not currently Py3 compatible. I am looking into migrating to the
"jsonpath" library as a longer term solution.

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm committed:
Merge branch 'master' into nh/omrs/atom

kaapstorm commented:
Closing. I broke this PR into several smaller chunks, and they have now all been merged, along with a few more tests and some documentation.
